### ISSUE TITLE
Log File Locking - Windows

### ISSUE THREAD

**Issue by jjrbg** [wazuh/wazuh#6128]
I've identified a problem with the agent on Windows locking log files it's monitoring so they cannot be moved or deleted. It's as if it keeps the file open and never closes it. This has a big impact if you have an archiving/maintenance script to compress, move and delete the logs each night (for example). As the files will just build up until disk space runs out.

It is quite simple to reproduce:

1. Create a test file to monitor e.g. C:\Test\test.log
2. Configure the agent to collect logs from localfile C:\Test\*.log
3. Using Process Explorer, you can see the file handle on ossec-agent.exe and Process Monitor shows the file being read approx every 2 seconds.
4. Try to delete the file (either del in cmd or Remove-Object in PowerShell) - this results in "Access Denied" or "File Does Not Exist". The file remains in place...

![Screenshot from 2020-09-29 10-28-44](https://user-images.githubusercontent.com/17181060/94540974-f49ba100-023e-11eb-9d94-30fff74b0dad.png)

Stopping the Wazuh service or killing the ossec-agent.exe process, releases the lock on the file, and it can then be deleted (or moved). As such, a possible workaround is to incorporate stopping the service as part of the nightly archiving process and restarting it once complete. But I don't think this is an ideal solution. 

Does anyone know if this behavior can be changed? It feels like a fairly significant issue. The Wazuh agent shouldn't interfere with the log files it's monitoring.

James

**Comment by antoniomanuelfr** [comment]
Hi @jjrbg 
Could you please tell us the version of the Wazuh agent that you are using? 
This issue was solved in 3.13 version (https://github.com/wazuh/wazuh/issues/4920) so it should be solved. 
I'm going to test it anyway and I reply to you ASAP
Kind regards.

**Comment by jjrbg** [comment]
Hi Antonio,
This is with the latest version of the agent. So 3.12-2.

Agent ossec.conf is default. /shared/default/agent.conf I'm pushing out from the manager includes:

```
<agent_config name="jamesg-iis">
  <localfile>
    <location>C:\Test\*.log</location>
    <log_format>syslog</log_format>
  </localfile>
</agent_config>
```

**Comment by jjrbg** [comment]
I've noticed the file delete action seems to queue for a while. Sometimes it can process it quite quickly, other times I've seen the file remains there for 1min+. As if the OS is waiting for the wazuh process to release the lock on the file.

**Comment by antoniomanuelfr** [comment]
Hi @jjrbg. 
My bad, I read wrongly your problem and I thought that it was the same issue. Sorry about that.

I have tried it using Powershell and you are right, logcollector takes some time in order to free the handle, but after some seconds, logcollector frees the handle and the file is properly removed.

Fortunally, there is an internal option called `logcollector.vcheck_files` that may help you. With this option you can choose the time that logcollector takes in order to check the files that it is monitoring. The default value is 64 seconds,  You can find more info [here ](https://documentation.wazuh.com/3.13/user-manual/reference/internal-options.html?highlight=internal#logcollector). 
You need to keep in mind that this may increase the resources that logcollector needs, be carefull setting that option.

Another way can be adding an sleep of 64 second into the script in order to give logcollector the time to free the handle. 
Also, you can see in the ossec.log file this message  that appears when logcollector frees the handle of the file that was deleted: 
```
2020/09/29 16:09:19 ossec-agent: ERROR: (1117): Error handling file 'C:\testlog\test2.log' (date).
2020/09/29 16:09:19 ossec-agent: INFO: (1959): File 'C:\testlog\test2.log' no longer exists.
```

Hope that this information is useful. 
Kind regards.

**Comment by jjrbg** [comment]
Hi @antoniomanuelfr 
Actually, I think you accidentally discovered the fix!... I think this is being caused by the duplicate file descriptor bug fixed in #4920. The files can/do remain there indefiniately after being deleted, not only for up to 64 seconds. Or at least until the process is killed.

I suppose that makes sense, unless both handles on the file are freed at the same time, it will always be locked by one of them? So the delete will never be allowed.

I mistakenly thoughts v3.12 was the latest, not v3.13. Apologies. I've reproduced the behavior on v3.13 of the agent, and although the files still "linger" around for a while, they are freed eventually. Which was different to what was happening in v3.12.

Thanks for your help!

**Comment by antoniomanuelfr** [comment]
Hi @jjrbg 
As you mentioned, the duplicate descriptor was the problem, logcollector left one opened handle and that was the reason why the file couldn't be removed.

If you feel like logcollector takes too long to free the handle, you can lower the value of the internal option that I shared previously.
We have tests that look for file descriptor leaks (in fact [#4920](https://github.com/wazuh/wazuh/issues/4920) was detected by one of those tests). 

If you experience an error like this, don't hesitate to contact us and we will check it as soon as we can. 

I will close this issue, as I think that this was solved. Feel free to re-open it if it's necesary.

Kind regards.
Antonio.
